tools/crm_mon.c

```diff
@@ -2994,60 +2994,41 @@ print_failed_actions(FILE *stream, pe_working_set_t *data_set)
 static stonith_history_t *
 reduce_stonith_history(stonith_history_t *history)
 {
-    stonith_history_t *new = NULL, *hp, *np, *tmp;
+    stonith_history_t *new = NULL, *hp, *np;
 
     for (hp = history; hp; ) {
-        for (np = new; np; np = np->next) {
-            if ((hp->state == st_done) || (hp->state == st_failed)) {
-                /* action not in progress */
-                if (safe_str_eq(hp->target, np->target) &&
-                    safe_str_eq(hp->action, np->action) &&
-                    (hp->state == np->state)) {
-                    if ((hp->state == st_done) ||
-                        safe_str_eq(hp->delegate, np->delegate)) {
-                        /* replace or purge */
-                        if (hp->completed < np->completed) {
+        stonith_history_t *hp_next = hp->next;
+
+        hp->next = NULL;
+        if (!new) {
+            new = hp;
+        } else {
+            for (np = new; ; np = np->next) {
```

Comment:
At this point, hp is the last item in the new list (since hp->next was set to NULL).

Since the list is sorted by newest first, I don't think we need to loop through the entire new list. We just need to compare the hp with the item before it.

Reply:
> At this point, hp is the last item in the new list (since hp->next was set to NULL).

Certainly the first loop is the last item.

> 
> Since the list is sorted by newest first, I don't think we need to loop through the entire new list. We just need to compare the hp with the item before it.

For example, hp should be compared against all new lists, considering that the results of different delegates are set to the new list.